ext/ddtrace_profiling_native_extension/ruby_helpers.c

```diff
@@ -22,3 +25,53 @@ void raise_unexpected_type(
     )
   );
 }
+
+#define MAX_RAISE_MESSAGE_SIZE 256
+
+struct raise_arguments {
+  VALUE exception_class;
+  char exception_message[MAX_RAISE_MESSAGE_SIZE];
+};
+
+static void *trigger_raise(void *raise_arguments) {
+  struct raise_arguments *args = (struct raise_arguments *) raise_arguments;
+  rb_raise(args->exception_class, "%s", args->exception_message);
+}
+
+void grab_gvl_and_raise(VALUE exception_class, const char *format_string, ...) {
+  struct raise_arguments args;
+
+  args.exception_class = exception_class;
+
+  va_list format_string_arguments;
+  va_start(format_string_arguments, format_string);
+  vsnprintf(args.exception_message, MAX_RAISE_MESSAGE_SIZE, format_string, format_string_arguments);
+
+  rb_thread_call_with_gvl(trigger_raise, &args);
```

Comment:
A general question: is it possible that it is not valid to acquire the GVL here? Is there a failure scenario like "cannot acquired GVL because ..." or worse a crash (that's caused by `rb_thread_call_with_gvl` itself, not what's in `trigger_raise`).

Reply:
`rb_thread_call_with_gvl` I believe does the correct validations: https://github.com/ruby/ruby/blob/d0888f7784d2eec575d342e757f8ad29deb39b70/thread.c#L1756-L1779

...although because this is a really awkward error to have, the VM will chose to die in those situations; looking at the comments the Ruby developers seemed to be a bit unsure of what to do with a failure in that situation.

But you got me thinking that it's harmless for us to add an additional test to check if the helper is being misused and handle that situation without letting Ruby crash the VM. Added in a45ba5ded51799d994dd6bd72952f6a1a8db7353 .

The "not a Ruby thread" situation I don't think it's worth guarding against explicitly because a) we never create non-Ruby threads, b) we guard against running any code on the signal handlers and c) it's really hard to do something sane in that situation (because the caller is expecting to raise an exception and stop execution -- we can't just return to the caller and keep executing whatever else comes after the code that wanted to raise an exception...).

Reply:
I like the changes in https://github.com/DataDog/dd-trace-rb/commit/a45ba5ded51799d994dd6bd72952f6a1a8db7353 👍 👍 👍 .
I agree with not guarding against `rb_thread_call_with_gvl` being called by non-Ruby C threads.